[AutoDiff] Progress/cleanup towards indirect passing support.

dan-zheng opened the pull request:
- `PrimalGenCloner::visitApplyInst` now checks the activity of indirect
  results, in addition to direct results.
- Remove non-general assertions (e.g. asserting that types are nominal types,
  which isn't true with generic type parameter types).
- Minor edits/cleanup.

dan-zheng committed:
[AutoDiff] Progress/cleanup towards indirect passing support.

- `PrimalGenCloner::visitApplyInst` now checks the activity of indirect
  results, in addition to direct results.
- Remove non-general assertions (e.g. asserting that types are nominal types,
  which isn't true with generic type parameter types).
- Minor edits/cleanup.

rxwei commented on lib/SIL/SILLocation.cpp:
```diff
@@ -217,6 +217,8 @@ MandatoryInlinedLocation::getMandatoryInlinedLocation(SILLocation L) {
   if (L.isInTopLevel())
     return MandatoryInlinedLocation::getModuleLocation(L.getSpecialFlags());
 
+  llvm::errs() << "FAIL SILLocation kind " << (unsigned)L.getKind() << "\n";
```
Revert this

rxwei commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -3174,6 +3174,13 @@ static ManagedValue createThunk(SILGenFunction &SGF,
                    outputOrigType,
                    outputSubstType);
   }
+  llvm::errs() << "CREATING A REABSTRACTION THUNK!\n";
```
Revert this also

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2380,9 +2382,15 @@ class PrimalGenCloner final
+    // Special handling logic only applies when `apply` has active results. If
+    // not, just do standard cloning.
+    SmallVector<SILValue, 4> allResults;
+    allResults.push_back(ai);
```
`ai` could be a tuple of multiple results. Use `extractElements` first.

dan-zheng replied:
I initially tried `extractElements`, but I changed it to this when I found `extractElements` was problematic. I don't believe activity analysis destructures tuple results, it treats them as one value.

rxwei replied:
Ah, sorry i didn't mean `extractElements`, but `collectAllActualResultsInTypeOrder`. `extractElements` would definitely not work because it generates new instructions which will not have any activity information (they wouldn't even be in the original function).

dan-zheng replied:
I see. I think `collectAllActualResultsInTypeOrder` expects an `ArrayRef<SILValue>` of tuple-extracted direct results though, based on its current caller and some print debugging.

```cpp
    // Extract all results from `pullbackCall`.
    SmallVector<SILValue, 8> dirResults;
    extractAllElements(pullbackCall, builder, dirResults); // <--- call to `extractAllElements`
    // Get all results in type-defined order.
    SmallVector<SILValue, 8> allResults;
    collectAllActualResultsInTypeOrder(
        pullbackCall, dirResults, pullbackCall->getIndirectSILResults(),
        allResults);
```

So I think what we want is to do sth like `collectAllActualResultsInTypeOrder` (to collect indirect results), except that if any `SILResultInfo` is direct, then we append `ai` (the entire direct result) itself?

rxwei replied:
Yes!

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3174,9 +3183,11 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    if (seed->getType().isReferenceCounted(getModule())) {
+      auto *rvi = builder.createRetainValue(origRetLoc, seed,
+                                            builder.getDefaultAtomicity());
+      assert(rvi->getParent() == getAdjoint().getEntryBlock());
```
K, i accidentally checked in this assertion. Please help remove this and the auto variable :) Thanks!

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22472/commits/4f369a758688637ee7d888acd64d14ecbfab96b6.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3174,9 +3183,11 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-    auto *rvi = builder.createRetainValue(origRetLoc, seed,
-                                          builder.getDefaultAtomicity());
-    assert(rvi->getParent() == getAdjoint().getEntryBlock());
+    if (seed->getType().isReferenceCounted(getModule())) {
```
WIP: All `AutoDiff` tests pass, but `TensorFlowRuntime/{tensor,model}_autodiff_runtime.swift` crash at runtime with this `seed->getType().isReferenceCounted(getModule())` condition.

dan-zheng replied:
Resolved in https://github.com/apple/swift/pull/22472/commits/4f369a758688637ee7d888acd64d14ecbfab96b6 by removing the `isReferenceCounted` condition and either emitting `retain_value` or `retain_value_addr`.

dan-zheng committed:
Fix refcounting logic.

- The `seed.getType()->isReferenceCounted` condition was problematic
  and let to crashes/test failures.
  - The new logic unconditionally retains. It checks whether the seed type
    is an address, and either emits `retain_value` or `retain_value_addr`.
    Cleanup does a similar check.
- Remove other `isReferenceCounted` condition, as it might cause problems.

Address feedback by @rxwei.

dan-zheng committed:
Fix seed adjoint initialization.

- Seed adjoint should be initialized differently based on whether the seed
  type is a buffer (use `setAdjointBuffer`) or value (use
  `initializeAdjointValue`).
  - Inline definition of `prepareSeedAdjoint` for the "value seed" case.
  - Both cases retain the seed and release the seed during cleanup.
- Re-add `retain_value` check to `test/AutoDiff/refcounting.swift`.

dan-zheng commented:
@swift-ci Please test tensorflow
